### ISSUE TITLE
setuptools build_ext always tries to load Cython

### ISSUE THREAD

**Issue by ax3l** [spack/spack#6558]
`spack install py-matplotlib` with
- mpl 2.0.2
- python 2.7.13
- GCC 6.3.0

(- cython 0.25.2)

fails on a Cython error in Debian 9.2 (note that cython is not a direct dependency of MPL).

### Expected Result

It should build & install :)

### Actual Result

```
==> Executing phase: 'build'
==> '/home/axel/src/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/python-2.7.13-rqdcehaqxz5gp2uplit25gtv6jm4g7fs/bin/python2.7' 'setup.py' '--no-user-cfg' 'build'
============================================================================
Edit setup.cfg to change the build options

BUILDING MATPLOTLIB
            matplotlib: yes [2.0.2]
                python: yes [2.7.13 (default, Dec  2 2017, 14:03:34)  [GCC
                        6.3.0 20170516]]
              platform: yes [linux2]

REQUIRED DEPENDENCIES AND EXTENSIONS
                 numpy: yes [version 1.13.3]
                   six: yes [using six version 1.10.0]
              dateutil: yes [using dateutil version 2.5.2]
           functools32: yes [using functools32]
          subprocess32: yes [using subprocess32]
                  pytz: yes [using pytz version 2017.2]
                cycler: yes [using cycler version 0.10.0]
               tornado: yes [using tornado version 4.3]
             pyparsing: yes [using pyparsing version 2.2.0]
                libagg: yes [pkg-config information for 'libagg' could not
                        be found. Using local copy.]
              freetype: yes [version 2.7.1]
                   png: yes [version 1.6.29]
                 qhull: yes [pkg-config information for 'qhull' could not be
                        found. Using local copy.]

OPTIONAL SUBPACKAGES
           sample_data: yes [installing]
              toolkits: yes [installing]
                 tests: no  [skipping due to configuration]
        toolkits_tests: no  [skipping due to configuration]

OPTIONAL BACKEND EXTENSIONS
                macosx: no  [Mac OS-X only]
                qt5agg: no  [PyQt5 not found]
                qt4agg: no  [PySide not found; PyQt4 not found]
               gtk3agg: no  [Requires pygobject to be installed.]
             gtk3cairo: no  [Requires cairocffi or pycairo to be installed.]
                gtkagg: no  [Requires pygtk]
                 tkagg: yes [installing; run-time loading from Python Tcl /
                        Tk]
                 wxagg: no  [requires wxPython]
                   gtk: no  [Requires pygtk]
                   agg: yes [installing]
                 cairo: no  [cairocffi or pycairo not found]
             windowing: no  [Microsoft Windows only]

OPTIONAL LATEX DEPENDENCIES
                dvipng: yes [version 1.14]
           ghostscript: yes [version 9.20]
                 latex: yes [version 3.14159265]
               pdftops: yes [version 0.48.0]

OPTIONAL PACKAGE DATA
                  dlls: no  [skipping due to configuration]

running build
running build_py
copying lib/matplotlib/mpl-data/matplotlibrc -> build/lib.linux-x86_64-2.7/matplotlib/mpl-data
UPDATING build/lib.linux-x86_64-2.7/matplotlib/_version.py
set build/lib.linux-x86_64-2.7/matplotlib/_version.py to '2.0.2'
running build_ext
failed to import Cython: /home/axel/.local/lib/python2.7/site-packages/Cython/Compiler/Scanning.so: undefined symbol: PyUnicodeUCS4_DecodeUTF8
error: Cython does not appear to be installed
```

### Steps to reproduce the issue

```console
$ spack install py-matplotlib
```

### Information on your system

- Debian 9.2 (x86_64)
- spack `develop` as of 80048ae4a

### Misc

Could this be related to the way python was build as in https://stackoverflow.com/a/2584554/2719194 ? Python seems to be indeed installed with `+pic+shared~tk~ucs4` when I check.

Running
```console
$ spack install py-matplotlib ^python+ucs4
```
then fails on
```
Cython/Compiler/Code.so: undefined symbol: PyFPE_jbuf
```

**Comment by ax3l** [comment]
If I add `depends_on('py-cython')` to the `py-matplotlib` package I get it compiled with and without `usc4` support.

**Comment by ax3l** [comment]
the cython dependency is pulled from `pdftops`

**Comment by tacaswell** [comment]
Are you building numpy from scratch?  That does requires cython.

Both `latex` and `pdftops` are used by just by calling them as subprocesses and should not involve cython either.

Can you get any more verbose logging of what it is doing when it fails?

ext here is 'external' not 'extension' I think.

**Comment by ax3l** [comment]
thx for weighting in!
Numpy and other dependencies are pre-build in that scenario (from scratch) and already in-place when mpl is getting built.

I added a little debug output to mpl's `setup.py` in
```python
class BuildExtraLibraries(BuildExtCommand):
    def run(self):
        for package in good_packages:
            print("starting with %s" % package)
            package.do_custom_build()
```

and it fails with:
```
running build_ext
starting with <setupext.Matplotlib object at 0x7f830241e950>
starting with <setupext.Python object at 0x7f830241e890>
starting with <setupext.Platform object at 0x7f8302b8cbd0>
starting with <setupext.Numpy object at 0x7f8302425c10>
starting with <setupext.Six object at 0x7f8302425a10>
starting with <setupext.Dateutil object at 0x7f8302425b50>
starting with <setupext.FuncTools32 object at 0x7f8302425cd0>
starting with <setupext.Subprocess32 object at 0x7f8302425d90>
starting with <setupext.Pytz object at 0x7f8302425dd0>
starting with <setupext.Cycler object at 0x7f8302425f50>
starting with <setupext.Tornado object at 0x7f8302425f90>
starting with <setupext.Pyparsing object at 0x7f830242e390>
starting with <setupext.LibAgg object at 0x7f830242ed50>
starting with <setupext.FreeType object at 0x7f830242eed0>
starting with <setupext.FT2Font object at 0x7f830243b510>
starting with <setupext.Png object at 0x7f830243b0d0>
starting with <setupext.Qhull object at 0x7f830243b210>
starting with <setupext.Image object at 0x7f830243b290>
starting with <setupext.TTConv object at 0x7f83023f6750>
starting with <setupext.Path object at 0x7f83023f6790>
starting with <setupext.ContourLegacy object at 0x7f83023f67d0>
starting with <setupext.Contour object at 0x7f83023f6810>
starting with <setupext.Delaunay object at 0x7f83023f6850>
starting with <setupext.QhullWrap object at 0x7f83023f6890>
starting with <setupext.Tri object at 0x7f83023f68d0>
starting with <setupext.SampleData object at 0x7f83023f6910>
starting with <setupext.Toolkits object at 0x7f83023f6950>
starting with <setupext.BackendTkAgg object at 0x7f83023f6b90>
starting with <setupext.BackendAgg object at 0x7f83023f6c50>
starting with <setupext.DviPng object at 0x7f83023f6d10>
starting with <setupext.Ghostscript object at 0x7f83023f6d50>
starting with <setupext.LaTeX object at 0x7f83023f6d90>
starting with <setupext.PdfToPs object at 0x7f83023f6dd0>
failed to import Cython: /home/axel/.local/lib/python2.7/site-packages/Cython/Compiler/Scanning.so: undefined symbol: PyUnicodeUCS4_DecodeUTF8
error: Cython does not appear to be installed
```

(I do explicitly not load a compatible Cython since it's not a MPL dependency but it needs one at build time - falling back to my system libs).

**Comment by ax3l** [comment]
Ah, actually the cython issue comes from after that loop, probably during `BuildExtCommand.run(self)`

**Comment by ax3l** [comment]
If I do not append ext. modules into the `ext_modules` by commenting the two lines out, Cython is not pulled in and it passes the build & install phases.

**Comment by tacaswell** [comment]
Starting from a clean conda enviroment I can build it cleanly with out cython:

```
21:44 $ conda create -p /tmp/py27 python=2.7 numpy pkgconfig libpng
Fetching package metadata .........
Solving package specifications: .

Package plan for installation in environment /tmp/py27:

The following NEW packages will be INSTALLED:

    certifi:    2016.2.28-py27_0 defaults
    libpng:     1.6.30-1         defaults
    mkl:        2017.0.3-0       defaults
    numpy:      1.13.1-py27_0    defaults
    openssl:    1.0.2l-0         defaults
    pip:        9.0.1-py27_1     defaults
    pkgconfig:  1.1.0-py27_0     defaults
    python:     2.7.13-0         defaults
    readline:   6.2-2            defaults
    setuptools: 36.4.0-py27_1    defaults
    sqlite:     3.13.0-0         defaults
    tk:         8.5.18-0         defaults
    wheel:      0.29.0-py27_0    defaults
    zlib:       1.2.11-0         defaults

Proceed ([y]/n)? y

#
# To activate this environment, use:
# > source activate /tmp/py27
#
# To deactivate an active environment, use:
# > source deactivate
#

✔ ~/source/p/matplotlib [v2.0.2|⚑ 3] 
21:45 $ source activate /tmp/py27
(py27) ✔ ~/source/p/matplotlib [v2.0.2|⚑ 3] 
21:45 $ conda list
# packages in environment at /tmp/py27:
#
certifi                   2016.2.28                py27_0    defaults
libpng                    1.6.30                        1    defaults
mkl                       2017.0.3                      0    defaults
numpy                     1.13.1                   py27_0    defaults
openssl                   1.0.2l                        0    defaults
pip                       9.0.1                    py27_1    defaults
pkgconfig                 1.1.0                    py27_0    defaults
python                    2.7.13                        0    defaults
readline                  6.2                           2    defaults
setuptools                36.4.0                   py27_1    defaults
sqlite                    3.13.0                        0    defaults
tk                        8.5.18                        0    defaults
wheel                     0.29.0                   py27_0    defaults
zlib                      1.2.11                        0    defaults
(py27) ✔ ~/source/p/matplotlib [v2.0.2|⚑ 3] 
21:45 $ cython
bash: cython: command not found
(py27) ✘-127 ~/source/p/matplotlib [v2.0.2|⚑ 3] 
21:45 $ git clean -xfd
(py27) ✔ ~/source/p/matplotlib [v2.0.2|⚑ 3] 
21:45 $ pip install .
Processing /home/tcaswell/source/p/matplotlib
Requirement already satisfied: numpy>=1.7.1 in /tmp/py27/lib/python2.7/site-packages (from matplotlib==2.0.2)
Collecting six>=1.10 (from matplotlib==2.0.2)
  Using cached six-1.11.0-py2.py3-none-any.whl
Collecting python-dateutil (from matplotlib==2.0.2)
  Using cached python_dateutil-2.6.1-py2.py3-none-any.whl
Collecting functools32 (from matplotlib==2.0.2)
Collecting subprocess32 (from matplotlib==2.0.2)
Collecting pytz (from matplotlib==2.0.2)
  Using cached pytz-2017.3-py2.py3-none-any.whl
Collecting cycler>=0.10 (from matplotlib==2.0.2)
  Using cached cycler-0.10.0-py2.py3-none-any.whl
Collecting pyparsing!=2.0.0,!=2.0.4,!=2.1.2,!=2.1.6,>=1.5.6 (from matplotlib==2.0.2)
  Using cached pyparsing-2.2.0-py2.py3-none-any.whl
Installing collected packages: six, python-dateutil, functools32, subprocess32, pytz, cycler, pyparsing, matplotlib
  Running setup.py install for matplotlib ... done
Successfully installed cycler-0.10.0 functools32-3.2.3.post2 matplotlib-2.0.2 pyparsing-2.2.0 python-dateutil-2.6.1 pytz-2017.3 six-1.11.0 subprocess32-3.2.7
(py27) ✔ ~/source/p/matplotlib [v2.0.2|⚑ 3] 
21:46 $ 
```

which makes me very confident that Matplotlib does not directly or indirectly depend on having cython available at build time.

**Comment by ax3l** [comment]
Ah, I see where it comes from. [Setuptools](https://github.com/pypa/setuptools/blob/v38.2.4/setuptools/command/build_ext.py#L17) pulls in on `from setuptools.command.build_ext import build_ext as BuildExtCommand` *any* Cython it can find:

```
try:
    # Attempt to use Cython for building extensions, if available
    from Cython.Distutils.build_ext import build_ext as _build_ext
except ImportError:
_build_ext = _du_build_ext
```

**Comment by ax3l** [comment]
@tacaswell thank you for your help, it's indeed a `setuptools` issue!

**Comment by ax3l** [comment]
The proper spack mitigation for the auto-enabled `cython` dependency for `setuptools` could be:
```
diff --git a/var/spack/repos/builtin/packages/py-setuptools/package.py b/var/spack/repos/builtin/packages/py-setuptools/package.py
index de7b47367..734295f3c 100644
--- a/var/spack/repos/builtin/packages/py-setuptools/package.py
+++ b/var/spack/repos/builtin/packages/py-setuptools/package.py
@@ -57,3 +57,4 @@ class PySetuptools(PythonPackage):
     depends_on('py-packaging@16.8:', when='@34.0.0:', type=('build', 'run'))
     depends_on('py-six@1.6.0:',      when='@34.0.0:', type=('build', 'run'))
     depends_on('py-appdirs@1.4.0:',  when='@34.0.0:', type=('build', 'run'))
+    depends_on('py-cython', type=('build', 'run'))
```

**Comment by ax3l** [comment]
Fixed with #7065
